Migrate to ESPHome External Component

cadwal opened the pull request:
This is my take on migrating to being an External Component instead of a Custom Component as that got removed in ESPHome 25.2.0. HDLC untested as before and the ASCII paths "works for me"...

While I prefer this way of defining the possible sensors and some other things about it, I would say that the state machine, the overall c++ usage and some other things are better in https://github.com/Beaky2000/esphome-p1mini by @Beaky2000.  
(LTNS Johnny)

cadwal committed:
External Component

MrGlad8 commented:
Hey, where can I find the new code to use?

cadwal commented:
> Hey, where can I find the new code to use?

It is in this branch: https://github.com/cadwal/esphome-p1reader/tree/main or the version from @Beaky2000 mentioned above.

NOTE: HDLC is definitely reported as broken in this code.

MrGlad8 commented:
> > Hey, where can I find the new code to use?
> 
> It is in this branch: https://github.com/cadwal/esphome-p1reader/tree/main or the version from @Beaky2000 mentioned above.
> 
> NOTE: HDLC is definitely reported as broken in this code.

Thank you, should I just copy the code right away? Should I leave the p1reader.h file as it is?

MrGlad8 commented:
> p1reader.h is no longer used in any way.
> 
> You need to put the esphome-components folder with contents next to your p1reader.yaml and then update p1reader.yaml to match the changes in the sample while preserving your uart config and wlan config. Also make sure that external_components: -source has a relative/absolute folder reference that matches your layout.

Thank you! But the only components I had before was this p1reader.h and the yaml file. So if the p1reader.h isnt used anymore what do I need other than the yaml-file? Should I exchange the p1reader.h with some other files before sending the yaml to my Slimmelezer? Sorry, this is above my  knowledge about this! Appreciate all help!

cadwal commented:
> > p1reader.h is no longer used in any way.
> > You need to put the esphome-components folder with contents next to your p1reader.yaml and then update p1reader.yaml to match the changes in the sample while preserving your uart config and wlan config. Also make sure that external_components: -source has a relative/absolute folder reference that matches your layout.
> 
> Thank you! But the only components I had before was this p1reader.h and the yaml file inside the config/esphome folder in my HomeAssistant. So if the p1reader.h isnt used anymore what do I need other than the yaml-file? Should I exchange the p1reader.h with some other files before sending the yaml to my Slimmelezer? Sorry, this is above my knowledge about this! Appreciate all help!

As I said, with the transformation to an external component, the p1reader.h file gets replaced by an entire folder structure with multiple files, called esphome_components in this branch. Then the yaml-file needs to be updated for that change and the other things, especially sensor stuff, that had to change when being a "real" EspHome component. There are a three different samples to look at for that.

MrGlad8 commented:
Thank you! I think I got it working, transferred the files to my esphome/esphome-components but now I wont get any data from my energy reader, its all N/A ...sigh... shouldnt have updated :/

cadwal commented:
> Thank you! I think I got it working, transferred the files to my esphome/esphome-components but now I wont get any data from my energy reader, its all N/A ...sigh... shouldnt have updated :/

Well, you did not have many options except not updating ESPHome at all and that would only have worked for a time.

Post your yaml-file (without passwords) and someone might be able to see what went wrong.

bortek commented:
Same situation here. Updated HA Core and it stopped working. If this will soon be merged then I can wait and test wafterwards. ping Pär, ska du kolla på det här ? @psvanstrom

Beaky2000 commented:
> ... If this will soon be merged then I can wait and test wafterwards. ...

Well, the last major PR took a little over a year to get approval, so I guess that is the current best estimate! :)

@psvanstrom, you should consider giving @cadwal access to maintain this project and/or offer to let him take over the whole thing. He seems to be the one doing all the work lately anyway!

cadwal committed:
Fix negative values, better time slices, HDLC

cadwal commented:
> > ... If this will soon be merged then I can wait and test wafterwards. ...
> 
> Well, the last major PR took a little over a year to get approval, so I guess that is the current best estimate! :)
> 
> @psvanstrom, you should consider giving @cadwal access to maintain this project and/or offer to let him take over the whole thing. He seems to be the one doing all the work lately anyway!

Or "we" as a sort of community could decide to go with your repo for this stuff? And/or perhaps even try to get an official component in the esphome repo? I had looked at that earlier but life...

cadwal commented:
WRT the latest push, there are some time slice fixes, the negative numbers and an HDLC path that has a better chance of working, the previous one was really broken...

@Beaky2000 If you want to OCD you can look at the HDLC scaling, perhaps even better if the divide by 1000.0 was configurable as it apparently differs between meters and when their firmware got "fixed".

bortek commented:
Well yes, go ahead and do it :) . I dont think we should expect this repo to be maintaned in the future, repo owner has probably lost interest in it 😒. I suggest we move on and use another repos which has been more active and have most of the new and good stuff in it. 

Is it this one ->?   https://github.com/Beaky2000/esphome-p1mini
Does it have the changes that are outlined in this PRor is it already working with 2025.2.5?

cadwal committed:
Time slice fixes

rodakswe commented:
So which one do we use now?
https://github.com/cadwal/esphome-p1reader/tree/main
or https://github.com/Beaky2000/esphome-p1mini
Just want as close functionality to this project as possible.

Beaky2000 commented:
> So which one do we use now? ...

At this time, the @cadwal one is closest to how this project works while mine is (intended to be) easiest for someone starting from scratch.

What the best way is going into the future requires some thinking about...

bortek commented:
I have modified yesterday my existing (psvatnstrom) config/code and updated it with cadwal's repo, mostly to not to risk historical data. It worked just fine.

cadwal commented:
> I have modified yesterday my existing (psvatnstrom) config/code and updated it with cadwal's repo, mostly not to risk loosing historical data. It worked just fine. So now am using cadwals codebase/config.

AFAICT it is all about keeping the sensors name property in the yaml the same as before.
